test-tap/test.js

```diff
@@ -746,6 +747,110 @@ test('timeout is refreshed on assert', t => {
 	});
 });
 
+test('teardown passing test', t => {
+	const teardown = sinon.spy();
+	return ava(a => {
+		a.teardown(teardown);
+		a.pass();
+	}).run().then(result => {
+		t.is(result.passed, true);
+		t.ok(teardown.calledOnce);
+	});
+});
+
+test('teardown failing test', t => {
+	const teardown = sinon.spy();
+	return ava(a => {
+		a.teardown(teardown);
+		a.fail();
+	}).run().then(result => {
+		t.is(result.passed, false);
+		t.ok(teardown.calledOnce);
+	});
+});
+
+test('teardown awaits promise', t => {
+	let tornDown = false;
+	const teardownPromise = delay(200).then(() => {
+		tornDown = true;
+	});
+	return ava(a => {
+		a.teardown(() => teardownPromise);
+		a.pass();
+	}).run().then(result => {
+		t.is(result.passed, true);
+		t.ok(tornDown);
+	});
+});
+
+test('teardowns run sequentially in order', t => {
+	const teardownA = sinon.stub().resolves(delay(200));
+	let resolveB;
+	const teardownB = sinon.stub().returns(new Promise(resolve => {
+		resolveB = resolve;
+	}));
+	return ava(a => {
+		a.teardown(() => teardownA().then(resolveB));
+		a.teardown(teardownB);
+		a.pass();
+	}).run().then(result => {
+		t.is(result.passed, true);
+		t.ok(teardownA.calledBefore(teardownB));
+	});
+});
+
+test('teardown with cb', t => {
+	const teardown = sinon.spy();
+	return ava.cb(a => {
+		a.teardown(teardown);
+		setTimeout(() => {
+			a.pass();
+			a.end();
+		});
+	}).run().then(result => {
+		t.is(result.passed, true);
+		t.ok(teardown.calledOnce);
+	});
+});
+
+test('teardown throws error', t => {
+	const teardown = sinon.stub().throws('TeardownError');
+	return ava(a => {
+		a.teardown(teardown);
+		a.pass();
+	}).run().then(result => {
+		t.is(result.passed, false);
+		t.is(result.error.name, 'TeardownError');
+		t.ok(teardown.calledOnce);
+	});
+});
+
+test('teardown throws error with previous error', t => {
+	const teardown = sinon.stub().throws('TeardownError');
+	return ava(a => {
+		a.teardown(teardown);
+		a.fail();
+	}).run().then(result => {
+		t.is(result.passed, false);
+		t.is(result.error.name, 'AssertionError');
+		t.ok(teardown.calledOnce);
+	});
+});
+
+test('only having assertion in teardown fails test', t => {
+	const teardown = sinon.spy();
+	return ava(a => {
+		a.teardown(() => {
+			teardown();
+			a.pass();
```

Comment:
This won't be allowed anyhow, so let's remove this test.

Reply:
Assuming you won't reply today, I'm gonna go ahead and replace it with a specific error test instead. If you don't approve, I'll remove it.

Reply:
Silly me. Forget what I said. There is already an error for that, as stated above...